### PR TITLE
UI extension for harvesting a record image

### PR DESCRIPTION
**Pull request by mirjamverloop**

### CODE REVIEW

File: europeana-image-harvester/index.html [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+          const processedAsset = await extension.space.waitUntilAssetProcessed(
+            rawAsset.sys.id,
+            'en-GB'
+          );
+
+          const publishedAsset = await extension.space.publishAsset(processedAsset);
+          await extension.entry.fields.heroImage.setValue({
+            sys: {
+              type: "Link",
```
Comment (monkemedia) [inline europeana-image-harvester/index.html]: Single quotes?

File: europeana-image-harvester/index.html [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        // get the image and attribution
+        async function harvestImageAndAttribution (recordUrl) {
+          const recordId = recordUrl.match('\/[0-9]+\/[a-zA-Z0-9_]+$')[0];
+          const apiKey = extension.parameters.installation.apiKey;
+
+          const response = await fetch(`https://api.europeana.eu/api/v2/record${recordId}.json?wskey=${apiKey}`);
+          const responseJson = await response.json();
+          if (response.ok) {
+            return { valid: true, json: responseJson, recordId: recordId };
+          } else {
```
Comment (monkemedia) [inline europeana-image-harvester/index.html]: Don't need an `else` here

File: europeana-image-harvester/index.html [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            });
+        }
+
+        // get the image and attribution
+        async function harvestImageAndAttribution (recordUrl) {
+          const recordId = recordUrl.match('\/[0-9]+\/[a-zA-Z0-9_]+$')[0];
+          const apiKey = extension.parameters.installation.apiKey;
+
+          const response = await fetch(`https://api.europeana.eu/api/v2/record${recordId}.json?wskey=${apiKey}`);
+          const responseJson = await response.json();
```
Comment (monkemedia) [inline europeana-image-harvester/index.html:58]: is .json method a promise too?

File: europeana-image-harvester/index.html [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            message: 'Enter a record page URL'
+          });
+          if (!recordUrl) return;
+
+          harvestImageAndAttribution(recordUrl)
+            .then(response => {
+              if (response.valid) {
+                clearFields();
+                populateFields(response.json, response.recordId);
+                document.getElementById('message').innerText = 'Processing';
```
Comment (monkemedia) [inline europeana-image-harvester/index.html]: Worth creating a `document.getElementById('message')` variable?

File: europeana-image-harvester/index.html [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+          extension.entry.fields.attribution.setValue(attribution.filter(Boolean).join(', '));
+          extension.entry.fields.rightsStatement.setValue(languageValue(providerAggregation.edmRights));
+
+          // create an asset
+          const asset = {
+            fields: {
+              title: {},
+              file: {}
+            }
+          }
```
Comment (monkemedia) [inline europeana-image-harvester/index.html]: Semicolon should go here. I personally don't use them, but I can see the project does :)

File: europeana-image-harvester/index.html [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+          asset.fields.title['en-GB'] = languageValue(providerProxy.dcTitle) ? languageValue(providerProxy.dcTitle) : recordId;
```
Comment (monkemedia) [inline europeana-image-harvester/index.html]: Wonder if it's worth putting the 'en-GB' into an enum?

**Review by monkemedia — COMMENTED** [review]
Fantastic work. Please feel free to argue my reviews

File: europeana-image-harvester/index.html [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+            return p.europeanaProxy === false;
+          });
+
+          const attribution = [languageValue(providerProxy.dcTitle),
+                               languageValue(providerProxy.dcCreator),
+                               languageValue(providerAggregation.edmDataProvider)];
+
+          extension.entry.fields.identifier.setValue(recordId);
+          extension.entry.fields.attribution.setValue(attribution.filter(Boolean).join(', '));
+          extension.entry.fields.rightsStatement.setValue(languageValue(providerAggregation.edmRights));
```
Comment (rwd) [inline europeana-image-harvester/index.html]: Individual web resources may have their own edm:rights which need to be favoured if present, else fallback to that on the aggregation.

File: europeana-image-harvester/index.html [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+            return r.about === image;
+          });
+
+          const proxies = response.object.proxies;
+          const providerProxy = proxies.find( p => {
+            return p.europeanaProxy === false;
+          });
+
+          const attribution = [languageValue(providerProxy.dcTitle),
+                               languageValue(providerProxy.dcCreator),
```
Comment (rwd) [inline europeana-image-harvester/index.html]: Individual web resources may have their own dc:creator which need to be favoured if present, else fallback to that on the proxy.

File: europeana-image-harvester/index.html [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+          // create an asset
+          const asset = {
+            fields: {
+              title: {},
+              file: {}
+            }
+          }
+          asset.fields.title['en-GB'] = languageValue(providerProxy.dcTitle) ? languageValue(providerProxy.dcTitle) : recordId;
+          asset.fields.file['en-GB'] = {
+            contentType: resource.ebucoreHasMimeType,
```
Comment (rwd) [inline europeana-image-harvester/index.html]: What happens if this is blank in the API response?

**Review by rwd — CHANGES_REQUESTED** [review]

File: europeana-image-harvester/index.html (hunk 3 above)
Comment (mirjamverloop) [reply europeana-image-harvester/index.html:58]: Yes, it is

File: europeana-image-harvester/index.html (hunk 4 above)
Comment (mirjamverloop) [reply europeana-image-harvester/index.html]: Good idea, I have created a new variable.

File: europeana-image-harvester/index.html (hunk 6 above)
Comment (mirjamverloop) [reply europeana-image-harvester/index.html]: I have created a new const variable: locale.

File: europeana-image-harvester/index.html (hunk 9 above)
Comment (mirjamverloop) [reply europeana-image-harvester/index.html]: There is a fallback now to 'application/octet-stream', which seems the most default one. Does that make sense? Should there be a check that it is an image?

File: europeana-image-harvester/index.html [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        extension.window.startAutoResizer();
+
+        const message = document.getElementById('message');
+        const locale = 'en-GB';
+
+        // show dialog to harvest image + attribution
+        async function showHarvestDialog() {
+          const recordUrl = await extension.dialogs.openPrompt({
+            title: 'Harvest',
+            message: 'Enter a record page URL'
```
Comment (rwd) [inline europeana-image-harvester/index.html]: Extend the message to: 'Enter a record page URL, or a record ID'

**Review by rwd — APPROVED** [review]

**Pull request merged**